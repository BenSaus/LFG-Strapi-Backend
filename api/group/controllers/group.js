"use strict";
const { sanitizeEntity } = require("strapi-utils");
const check = require("../../checkFunctions");

/**
 * Read the documentation (https://strapi.io/documentation/v3.x/concepts/controllers.html#core-controllers)
 * to customize this controller
 */

module.exports = {
    async removeMember(ctx) {
        const { groupId } = ctx.request.body;
        const memberIdToRemove = Number(ctx.request.body.memberId);
        const requestingUserId = ctx.state.user.id;

        // VALIDATE IDs

        const group = await strapi.services.group.findOne({
            id: groupId,
        });

        check.groupMustBeValid(group);
        check.userMustBeGroupLeader(group, requestingUserId);
        check.userMustBeGroupMember(group, memberIdToRemove);

        try {
            const updatedMembers = group.members.filter(
                // WARNING: Here I am assuming Id's are Numbers!!!
                (member) => member.id !== memberIdToRemove
            );

            const updatedGroup = {
                members: updatedMembers,
            };
            const entity = await strapi.services.group.update(
                { id: groupId },
                updatedGroup
            );

            const sanitizedGroup = sanitizeEntity(entity, {
                model: strapi.models.group,
            });

            return { group: sanitizedGroup };
        } catch (error) {
            check.throwInternalServerError(error);
        }
    },

    async leaveGroup(ctx) {
        const groupId = ctx.request.body.id;
        const requestingUserId = ctx.state.user.id;

        // VALIDATE ID

        // Verify this request is from a member of the group
        const group = await strapi.services.group.findOne({
            id: groupId,
        });

        check.groupMustBeValid(group);
        check.userMustBeGroupMember(group, requestingUserId);

        try {
            const updatedMembers = group.members.filter(
                (member) => member.id !== requestingUserId
            );

            const updatedGroup = { members: updatedMembers };

            const entity = await strapi.services.group.update(
                { id: groupId },
                updatedGroup
            );

            const sanitizedGroup = sanitizeEntity(entity, {
                model: strapi.models.group,
            });

            return { group: sanitizedGroup };
        } catch (error) {
            check.throwInternalServerError(error);
        }
    },

    // async createGroup(ctx) {}
    // async updateGroup(ctx) {
    //      leader only
    // }
    // async deleteGroup(ctx) {}
};
