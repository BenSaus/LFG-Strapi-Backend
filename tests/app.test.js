const { setupStrapi } = require("./helpers/strapi");
const strapiUser = require("./helpers/strapiUser");
const loadStrapiData = require("./helpers/strapiData");

beforeAll(async (done) => {
    await setupStrapi();
    const testUser = await strapiUser.createTestUser();
    await strapiUser.setAuthenticatedPermissions();
    await loadStrapiData(strapi, testUser);
    done();
});

it("strapi is defined", () => {
    expect(strapi).toBeDefined();
});

require("./user");
require("./group/group");
