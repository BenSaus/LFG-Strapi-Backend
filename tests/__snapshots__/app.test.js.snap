// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create Invite 1`] = `
Object {
  "data": Object {
    "createInvite": Object {
      "invite": Object {
        "group": Object {
          "id": "9",
          "name": "Test Wombats",
        },
        "id": "5",
        "invitee": Object {
          "id": "1",
          "username": "Jordan",
        },
        "message": "Heya invite here",
      },
    },
  },
}
`;

exports[`Create group 1`] = `
Object {
  "data": Object {
    "createGroup": Object {
      "group": Object {
        "booking_status": "notBooked",
        "id": "8",
        "max_age": 83,
        "member_max": 3,
        "min_age": 31,
        "name": "Test Wombats",
        "preferred_rooms": Array [],
      },
    },
  },
}
`;

exports[`Get Groups 1`] = `
Object {
  "data": Object {
    "groups": Array [
      Object {
        "id": "1",
        "name": "Flaming Giraffes",
      },
      Object {
        "id": "2",
        "name": "Rabid Raccoons",
      },
      Object {
        "id": "3",
        "name": "Diabetic Dingos",
      },
      Object {
        "id": "4",
        "name": "Bad Wombats",
      },
      Object {
        "id": "5",
        "name": "Misunderstood Mice",
      },
      Object {
        "id": "6",
        "name": "Argile Sweaters",
      },
      Object {
        "id": "7",
        "name": "Agitated Ardvarks",
      },
      Object {
        "id": "8",
        "name": "Test Wombats",
      },
    ],
  },
}
`;
